### ISSUE TITLE
XGBoost library version 1.4.2 is having performance issues compared to previous version libraries, verified against 1.2.1

### ISSUE THREAD

**Issue by SrikanthPusarla** [dmlc/xgboost#7410]
**Issue Description:**
XGBoost library version 1.4.2 is having below performance issues compared to previous version libraries when tested with same data and all other parameters, Verified this against 1.2.1 version.
1.	1.4.2 is resulting out larger model file size which is almost double to the model size generated by 1.2.1 version
2.	1.4.2 prediction call is taking quite longer time over 1.2.1 prediction call 

               
I made simple xgboost runs (with default parmeters) from  both 1.4.2 & 1.2.1 versions on sklearn data set of just 569 records on Databricks environment, The observations are:
1.	1.2.1 version generated model file of size **43690 bytes  (~ 43 KB)** whereas 1.4.2 version generated **82123 bytes (~ 82 KB)**. Which is **double the size** over 1.2.1
2.	1.4.2 prediction(predict_porba) call over single record took around **23879 micro seconds** whereas the same from 1.2.1 took around just **753 micro seconds** 

Please check the proof of code with results attached; Ran on Databricks 8.4 ML environment
•	XGBoost-1.4.2.html: code for 1.4.2 version
•	XGBoost-1.2.1.html: code for 1.2.1 version
[xgboost.zip](https://github.com/dmlc/xgboost/files/7510730/xgboost.zip)

**Comment by trivialfis** [comment]
For the size of pickle, that's because we switched to JSON as underlying serialization format.  The rationale behind is that we can't modify the binary format for extended features like categorical data support without breaking existing models.

As for the prediction time, we switched to `inplace_predict` by default for sklearn interface.  It bypass the construction of `DMatrix`, which saves time and memory.  But for a small dataset that runs in micro seconds, the logic for determining whether `inplace_predict` can be used become a bottleneck.

**Comment by trivialfis** [comment]
Thanks for sharing your results.  The pickle size gain is expected.  But as a mitigation, you can always use `save_model` with binary format instead of pickle, the `save_model` saves a stable model format while pickle itself is non-stable. 

For the prediction, you can bypass the sklearn model wrapper and run it directly with `cls.get_booster().inplace_predict` as a mitigation.  In general, for low latency prediction, you can choose one of those model deployment libraries like `treelite` or `fil` in `cuml` to name a few.  Most of the algorithms in XGBoost are designed/optimized to handle large batch of data to increase throughput, as a result we don't usually benchmark latency.

**Comment by trivialfis** [comment]
Feel free to reopen if there are unresolved issues.

**Comment by SrikanthPusarla** [comment]
Hi 
I tried your suggestions, 

1. saved the model by save_model into binary format (.bin) which resulted only **40KB** file whereas pickle generated **81KB**.
2. And for the inference time, tried get_booster().inplace_predict which took only **729 milli seconds** whereas predict_proba took **4007 milli seconds**

I have few other observations on the api you suggsted:  

- get_booster().inplace_predict  is giving only one prediction value i.e. for label 1
- There is difference in value for parameter tree_method for binary model loaded back,
-   o	Actual model trained:
XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1, colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1, importance_type='gain', interaction_constraints='', learning_rate=0.300000012, max_delta_step=0, max_depth=6, min_child_weight=1, missing=nan, monotone_constraints='()', n_estimators=100, n_jobs=16, num_parallel_tree=1, random_state=0, reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1, **tree_method='exact',** validate_parameters=1, verbosity=None)

- o	Binary Model loaded back:
XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1, colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1, importance_type='gain', interaction_constraints='', learning_rate=0.300000012, max_delta_step=0, max_depth=6, min_child_weight=1, missing=nan, monotone_constraints='()', n_estimators=100, n_jobs=16, num_parallel_tree=1, random_state=0, reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1, **tree_method='auto',** validate_parameters=1, verbosity=None)

But all the scores are matching between two.
Wont it be issue?

**Comment by trivialfis** [comment]
Hi, there are 2 prediction functions, `predict` and `inplace_predict` from `Booster`.  You can choose either one as you see fit.

> get_booster().inplace_predict is giving only one prediction value i.e. for label 1

For binary classification (`load_breast_cancer`), the prediction is the result from the sigmoid function, which is the probability of it being positive.  The `predict_proba` is just a wrapper over `inplace_predict` that transforms this into a 2 column array by appending `1 -  inplace_predict()` as the probability of negative labels.

> There is difference in value for parameter tree_method for binary model loaded back,

The default (as of xgboost 1.5) `tree_method` is `exact` for a small datasets like breast cancer.  You can specify the parameter explicitly if you don't want the configuration.

**Comment by SrikanthPusarla** [comment]
Yes, It worked with reading .bin model by passing the model parameters which were used initially to train instead of default initialization, then it is good.
[XGBoost-1.4.2-3.html.zip](https://github.com/dmlc/xgboost/files/7536623/XGBoost-1.4.2-3.html.zip)

Thanks
